building/trusted_applications.rst

```diff
@@ -422,7 +435,7 @@ or with pkcs11-tool using a Nitrokey HSM
      --input-file /tmp/hashtosign | \
      base64 > sigfile
 
-3. Manually (or with an extra make target) stitch the TA together using
+4. Manually stitch the TA and signature together
 
 .. code-block:: sh
```

Comment:
Instead of doing steps 2-4 we are using re-signing of TA elf binary.

Some notes of that method:
- No need to provided any metadata externally from build process -- they are already in elf binary.
- TA_PUBLIC_KEY and TA_SIGN_KEY must match in key size -- otherwise the signing can cause issues
  - We have RSA-4096 devel key external to OP-TEE OS source tree.
  - And then public key for release signing

Have not studied the TA signing format that deeply if key size shortcoming could be fixed in tooling.

In theory one could improve the `sign_encrypt.py` so that one can use PKCS11 modules directly (eg. handle the Nitrokey HSM signing with that). (python-pkcs11 could be your friend in that)

In theory the `sign_encrypt.py` could also support re-signing with support for PKCS11 to avoid proprietary implementations.

Reply:
> In theory one could improve the `sign_encrypt.py` so that one can use PKCS11 modules directly (eg. handle the Nitrokey HSM signing with that). (python-pkcs11 could be your friend in that)
> 
> In theory the `sign_encrypt.py` could also support re-signing with support for PKCS11 to avoid proprietary implementations.

I think we should keep this script as simple as possible for the sake of easy maintenance, auditing etc.
Btw., in some scenarios we are using airgapped machines for signing and only having the hash there keeps the implementations on that side simple and agnostic to the TA format.

Putting a note that TA_PUBLIC_KEY and TA_SIGN_KEY must have a compatible format and that one can re-sign the TAs with a customized `sign_encrypt.py` instead of passing hashes is a good idea.
No need to get further into the details, as people implementing offline-signing should be familiar with protocols and algorithms involved.

Reply:
> I think we should keep this script as simple as possible for the sake of easy maintenance, auditing etc.
Btw., in some scenarios we are using airgapped machines for signing and only having the hash there keeps the implementations on that side simple and agnostic to the TA format.

I am with @wamserma on this one. Theoretically we could extend the `sign_encrypted.py` script with PKCS11 features, however it would make the script complex and it would be very hard to cover a lot of use cases (e.g. in my case I have no access to the HSM hence PKCS11 implemented within `sign_encrypted.py` would be useless).

> Putting a note that TA_PUBLIC_KEY and TA_SIGN_KEY must have a compatible format

Aren't TA_PUBLIC_KEY and TA_SIGN_KEY supposed to be the **very same key**, i.e. TA_PUBLIC_KEY is actually exported from the TA_SIGN_KEY?

> that one can re-sign the TAs with a customized sign_encrypt.py instead of passing hashes is a good idea.

Yes this is a good idea, I'll add it to the documentation next week.

Reply:
> In theory one could improve the sign_encrypt.py so that one can use PKCS11 modules directly (eg. handle the Nitrokey HSM signing with that). (python-pkcs11 could be your friend in that)

If interested, I have something in this area. See the 3 top commits from https://github.com/etienne-lms/optee_os/commits/pkcs11-signed-tas:
* ta: mk: arm: CFG_SIGN_TA_WITH_PKCS11=y to sign TAs with PKCS#11 …
* scripts: get public key from PKCS#11 token …
* scripts: sign_encrypt.py: use keys from a PKCS#11 token …

Reply:
@wamserma @vesajaaskelainen I've just committed the changes discussed above.